src/main/java/teammates/storage/api/StudentsDb.java

```diff
@@ -407,45 +407,60 @@ public void updateStudent(String courseId, String email, String newName,
             String lastName = StringHelper.splitName(newName)[1];
             
             if (isEmailChanged) {
-                CourseStudent newCourseStudent = new CourseStudent(newEmail, newName, newGoogleId, newComments,
-                                                                   courseId, newTeamName, newSectionName);
-                newCourseStudent.setLastName(lastName);
-                newCourseStudent.setCreatedAt(courseStudent.getCreatedAt());
-                if (keepUpdateTimestamp) {
-                    newCourseStudent.setLastUpdate(courseStudent.getUpdatedAt());
-                }
-                
+                CourseStudent newCourseStudent = new CourseStudent(newEmail, newName, newGoogleId,
+                        newComments,
+                        courseId, newTeamName, newSectionName);
                 try {
-                    createStudent(new StudentAttributes(newCourseStudent), hasDocument);
+                    updateStudentChangeEmail(newCourseStudent, lastName, courseStudent, hasDocument,
+                            keepUpdateTimestamp, courseId, email);
                 } catch (EntityAlreadyExistsException e) {
                     StudentAttributes existingStudent = (StudentAttributes) e.existingEntity;
                     String error = ERROR_UPDATE_EMAIL_ALREADY_USED
                             + existingStudent.getName() + "/" + existingStudent.getEmail();
                     throw new InvalidParametersException(error);
                 }
-                
-                deleteStudent(courseId, email);
-                
             } else {
-                courseStudent.setName(newName);
-                courseStudent.setLastName(lastName);
-                courseStudent.setComments(newComments);
-                courseStudent.setGoogleId(newGoogleId);
-                courseStudent.setTeamName(newTeamName);
-                courseStudent.setSectionName(newSectionName);
-                
-                if (hasDocument) {
-                    putDocument(new StudentAttributes(courseStudent));
-                }
-            
-                // Set true to prevent changes to last update timestamp
-                courseStudent.keepUpdateTimestamp = keepUpdateTimestamp;
+                updateStudentDetails(newName, newTeamName, newSectionName, newEmail, newGoogleId,
+                        newComments, hasDocument, keepUpdateTimestamp, courseStudent, lastName);
             }
         }
         
         log.info(Const.SystemParams.COURSE_BACKUP_LOG_MSG + courseId);
         getPm().close();
     }
+    
+    public void updateStudentChangeEmail(CourseStudent newCourseStudent, String lastName,
+            CourseStudent courseStudent, boolean hasDocument, boolean keepUpdateTimestamp, String courseId,
+            String email)
+            throws InvalidParametersException, EntityAlreadyExistsException {
```

Comment:
As i said earlier this is the problem `New exception is thrown in catch block, original stack trace may be lost`